pkg/controller: fix incorrect backoff time for next retry

ArthurChiao opened the pull request:
It is observed that on agent start, controller "ipcache-inject-labels" will take several minutes or even longer to succeed, with status like,

```
$ cilium status --brief
error in controller ipcache-inject-labels: failed to inject labels into ipcache: k8s cache not fully synced

$ cilium status --all-controllers | grep "ipcache-inject-labels"
ipcache-inject-labels                         never          4m10s ago 331     failed to inject labels into ipcache: k8s cache not fully synced
```

It can be assured that k8s cache synchronizes quite fast, in about 1~2s.

Digging into the code, we found that the agent received much events (`NodeUpdated`), which will trigger one `ControllerUpdate` per events, and which in turn makes the controller loop increased retry count hundreds of times in several ms (during which period k8s cache is not ready), and when the event storm gone, the controller has to wait hundreds of seconds to hit `case runTimer.After(interval)`, although the k8s cache is already.  Related code in v1.11.10 (same logic in master): https://github.com/cilium/cilium/blob/v1.11.10/pkg/controller/controller.go#L266

This patch fixes the problem by distinguishing different retry types, and increasing backoff time only with the "timerExpired" type. Besides, detailed retry stats are also included into scoped log for better observability.

```release-note
Fix long-time failure of "ipcache-inject-labels" controller due to incorrect backoff time for retry
```

Signed-off-by: ArthurChiao <arthurchiao@hotmail.com>

pippolo84 commented on pkg/controller/controller.go:
```diff
@@ -197,8 +195,21 @@ func (c *Controller) runController() {
+		retryReasonControllerUpdated   = "controllerUpdated"
+		retryReasonControllerTriggered = "controllerTriggered"
+		retryReasonTimerExpired        = "timerExpired"
+	)
```
Nit: better declare these as constants and move them at line 23 with all the others.

ArthurChiao replied:
Going to move them, thanks! @pippolo84

pippolo84 reviewed (changes requested):
Nice, thank you! Just a nit to be addressed.

aanm commented on pkg/controller/controller.go:
```diff
@@ -279,23 +290,26 @@ func (c *Controller) runController() {
 			c.mutex.RUnlock()
 			runFunc = true
+			retriesByReason[retryReasonControllerUpdated]++
 
```
Wouldn't it solve the bug if we did a `errorRetries = 0` here?

ArthurChiao replied:
Hi @aanm, for the case I've met, we'll hit `case <-c.updated` hundreds of times before the first time we hit `case timer.After(interval)`, resetting `errorRetries` leaves the two cases into race conditions in `select` the after the next run, I'm not sure if the timer expire will always win (but certainly the wait time will dramatically be decreased before we succeed). Besides, I assume we should also reset it in the `case c.trigger`?

One consideration if changing in this way: the variable seems to lose its meanings if resetting it before we succeeded once yet; on the other hand, it'll be more helpful if we can distinguish these retries in logs or metrics, in which we'll know better about what the controller was doing.

christarazi reviewed (changes requested):
Thanks for the PR. I agree that I think this is a thundering herd problem. 

However, I think this fix is too specific to the use case and potentially too low-level of a fix for the actual problem. Let me explain my thought process. 

To give a bit of background, the `ipcache-inject-labels` controller was initially implemented for [`kube-apiserver` policy matching](https://github.com/cilium/cilium/pull/17823) (feature in v1.11). In order to plumb the kube-apiserver IP into the ipcache, we needed to be sure that the K8s caches have been synced so that we have the most up-to-date state before acting, hence the error you're talking about. More importantly, we want to avoid the situation where, due to having incomplete information of the cluster, we upsert transient identities into the ipcache, as this could lead to erroneous policy drops; the drops would resolve eventually, but we've opted for behavior where we'd rather wait for state to sync and avoid transient drops.

(Just for reference, there are two K8s watchers that trigger this `ipcache-inject-labels` controller because they might carry kube-apiserver-related state: K8s Endpoints, CiliumNode. See [diagram](https://github.com/cilium/cilium/blob/master/pkg/ipcache/metadata.go#L483-L506) for a visual.)

While this controller is getting triggered "too often" when a burst of node events come in, this approach fundamentally changes when the retry interval is calculated for **all** controllers. I think it's best to not change the behavior for all controllers, but rather fix it by **treating this controller as a special case**.

I can think of a few different approaches, which all attempt to mitigate the problem here, but without changing behavior for all controllers.

1. Set a max error retry duration (i.e. after error count reaches threshold, retry duration is hardcoded to X)
2. Set a fixed error retry duration (i.e. retry duration doesn't increase as more errors occur)
3. Allow the `ipcache-inject-labels` controller to only be invoked every Xms (i.e. something similar to `pkg/trigger`)

(3) requires us to know what should be X, and it's not obvious to me what that should be, as this will be a time window in which ipcache information will be artificially delayed. (1) and (2) are just different flavors of the same idea. I think either one is OK, not sure there's a difference.

ArthurChiao committed:
pkg/controller: introduce max retry interval for controllers

Add a MaxRetryInterval option to ControllerParams struct, which allows
users to configure the max error retry interval for each controller.
This can mitigate the thundering herd problem as reported in #21886.

Signed-off-by: ArthurChiao <arthurchiao@hotmail.com>

christarazi reviewed (approved):
Awesome, looks good, thanks!

pippolo84 reviewed (approved):
LGTM. Thanks :+1:
